### DISCUSSION TITLE
Proposal: allow ref enumeration over arrays

### DISCUSSION THREAD

**Discussion by huoyaoyuan** [dotnet/csharplang#8820] [Language Ideas]
### Proposed syntax:
```csharp
void M(X[] a)
{
    foreach (ref X x in a)
    {

    }
}
```

### Currently alternative:
```csharp
void M(X[] a)
{
    foreach (ref X x in a.AsSpan())
    {

    }
}
```

### Benifits:
Syntax completeness.
Can work if the `Span` type is not present.
**Since array enumeration is specially treated by the compiler, the compiler can generate simpler code to reduce the stress of JIT.** [Currently](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcDMACR2DC2A3utmbjnACzYCyAFAJYB2wA2gLrZgCUp5JNOWHYAZgHsATgFMwAYwAW9GaOwtgatc24A6AIIQAygAcwzejz5CRZQTZuNsAXmwAGANz8bAXy/Zf1mR+WLg0tAhMrJzcVsJ29hKS2JEajE4ejAA8YDoAMtLMAObACu6MANTlsfbx9uRgbIwc6Z6BIgHCAd5AA=) JIT generate worse code when using `AsSpan()`.

### Risk
No risk:
Array is specially treated by the compiler. This feature should not require any member defined in metadata.
This is about introducing new binding rule, not new syntax.
The semantic is straight.

**Comment by YairHalberstadt** [comment]
I feel like it would be a better idea to open an issue on dotnet/runtime to improve the Jit.

**Comment by reflectronic** [comment]
This is still a hole in the language. I opened [an issue about it](https://github.com/dotnet/csharplang/issues/3209) a while ago, and it appears that the feature is a little half baked.
